numpy/core/src/multiarray/descriptor.c

```diff
@@ -773,6 +773,63 @@ _is_tuple_of_integers(PyObject *obj)
 }
 
 /*
+ * helper function for _use_inherit to disallow dtypes of the form
+ * (old_dtype, new_dtype) where either of the dtypes contains python
+ * objects - these dtypes are not useful and can be a source of segfaults,
+ * when an attempt is made to interpret a python object as a different dtype
+ * or vice versa
+ * an exception is made for dtypes of the form ('O', [('name', 'O')]), which
+ * people have been using to add a field to an object array without fields
+ */
+static int
+validate_structured_object_dtype(PyArray_Descr *new, PyArray_Descr *conv)
+{
+    PyObject *names, *fields, *name, *tup;
+    PyArray_Descr *dtype;
+
+    if (!PyDataType_REFCHK(new) && !PyDataType_REFCHK(conv)) {
+        return 0;
+    }
+    if (PyDataType_HASFIELDS(new)) {
```

Comment:
I think by organizing the tests this way you are trying to allow cases like `np.dtype(('i4', 'O'))`. But I think we should rule that out too: Consider the strange fact that `np.dtype(('i8', 'O')).hasobject` is `True`.

Maybe reorganize the if-statements here a little, eg start with `if (new->kind != 'O')`,  then `if (!PyDataType_HASFIELDS(conv)) `, then `if (PyTuple_GET_SIZE(names) != 1)` an so on, an fail if any is true.